### ISSUE TITLE
Laravel MongoDB No suitable servers found ('serverSelectionTryOnce' set) PHP 5.6

### ISSUE THREAD

**Issue by bisrinursa** [mongodb/mongo-php-driver#667]
I try to connect laravel with mongodb but i've got error "no suitable servers found"

Here's my machine
```
Description:    Ubuntu 14.04.5 LTS
Release:        14.04
Codename:       trusty

PHP Version => 5.6.31-2+ubuntu14.04.1+deb.sury.org+1

System => Linux work-vbox 3.13.0-24-generic #47-Ubuntu SMP Fri May 2 23:30:00 UTC 2014 x86_64
Server API => Command Line Interface
Virtual Directory Support => disabled
Configuration File (php.ini) Path => /etc/php/5.6/cli
Loaded Configuration File => /etc/php/5.6/cli/php.ini

MongoDB support => enabled
MongoDB extension version => 1.2.11
MongoDB extension stability => stable
libbson bundled version => 1.5.5
libmongoc bundled version => 1.5.5
libmongoc SSL => enabled
libmongoc SSL library => OpenSSL
libmongoc crypto => enabled
libmongoc crypto library => libcrypto
libmongoc crypto system profile => disabled
libmongoc SASL => disabled
```

**Comment by bisrinursa** [comment]
Now, it's working when i remove `DB_REPLICA_SET` in config/database.php.

The code I cloned from the repo had it set to `env('DB_REPLICA_SET', NULL)` and works fine on them so I never suspected that it was the cause before this. Is this a known issue? version related perhaps?

**Comment by jmikola** [comment]
Searching GitHub for `DB_REPLICA_SET` turned up [leroy-merlin-br/mongolid-laravel](https://github.com/leroy-merlin-br/mongolid-laravel). Is that the library you are using to interface with the driver?

The MongoDB connection string supports a [`replicaSet` option](https://docs.mongodb.com/manual/reference/connection-string/#urioption.replicaSet), which should specify the name of the replica set to which you are connecting. [`MongoDB\Driver\Manager::__construct()`](http://php.net/manual/en/mongodb-driver-manager.construct.php) likewise accepts a `replicaSet` option in its URI options (which are merged into the first connection string argument). If you are connecting to a standalone mongod or a shard cluster, the `replicaSet` option should be omitted.

If you were connecting to a standalone mongod and a `replicaSet` option was specified, it's possible that you're hitting [this error](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#verifying-setname-with-topologytype-single). The presence of the option would instruct libmongoc (which we use in the driver) to assert that the server to which we're connecting does report the same replica set name in its `isMaster` response. In practice, that is likely a configuration error.

If you are connecting to a replica set, the option should match the name in the replica set configuration. The driver uses this name to ensure that it doesn't start communicating with hosts that may belong to another replica set (possible if IPs are re-used across different deployments, or if hostnames from two different replica sets end up in the same connection string).

If this is a case of [leroy-merlin-br/mongolid-laravel](https://github.com/leroy-merlin-br/mongolid-laravel) specifying a `replicaSet` option when it shouldn't, I'd encourage you to open a bug report on that library so this can be fixed upstream. Feel free to link to this thread for context.

**Comment by lindelius** [comment]
We had an issue related to the replica set option, as well, and I’m suspecting you have the same. You can no longer set it to `null` when you do not wish to connect to a replica set (which worked before) because the new driver requires it to be a string, if it’s set. Also, setting it to an empty string doesn’t work either as it then tries to connect to a replica set with an empty name. What we did is to create a helper method that we use in the `config/database.php` like this

```php
    'uriOptions' => getMongoDbUriOptions(),
```

**Comment by jmikola** [comment]
@lindelius: 1.3.0 did add stricter type checking for connection strings ([CDRIVER-782](jira.mongodb.org/browse/CDRIVER-782)) and values in the URI options array ([PHPC-887](https://jira.mongodb.org/browse/PHPC-887)), but I'm not sure that is affecting the user as they appear to be using 1.2.11.

If a PHP `null` was appended to the connection string, it would be an empty string. There would be no error, but libmongoc would indeed fail to match the replica set name during server discovery. I'd expect that to result in a server selection failure in any version of the PHP driver, as that behavior in libmongoc would not have changed recently. Generally, libmongoc only reports connection string errors if a query string parameter expects a boolean or integer and the substring doesn't parse correctly. By default, any option expecting a string should be valid.

If `null` was used in the URI options array, the 1.3.0 driver would indeed throw InvalidArgumentException upon constructing the Manager. I expect that may be the error that you experienced, but it's immediately apparently and a quick fix (vs. a cryptic server selection failure at runtime).

**Comment by bisrinursa** [comment]
@jmikola me and my team uses [jenssegers/laravel-mongodb](https://github.com/jenssegers/laravel-mongodb) so I think the problem is not there

@lindelius yes! my problem closely resembles yours, didn't thought of such a creative workaround though. And yes @jmikola I used `null` at first but since the InvalidArgumentException says that it expects a string, I entered empty string instead and ended up with the server selection failure.

Anyway, I'm closing this since I got it working, thanks for the help guys, cheers!

**Comment by jmikola** [comment]
> The code I cloned from the repo had it set to `env('DB_REPLICA_SET', NULL)`

@bisrinursa: What is "the repo" that you were referring to? I didn't find `DB_REPLICA_SET` mentioned anywhere in the [jenssegers/laravel-mongodb](https://github.com/jenssegers/laravel-mongodb) repository. The [Configuration](https://github.com/jenssegers/laravel-mongodb#configuration) section in the project's README only has the following example:

> You can connect to multiple servers or replica sets with the following configuration:
>
> ``` 
> 'mongodb' => [
>     'driver'   => 'mongodb',
>     'host'     => ['server1', 'server2'],
>     'port'     => env('DB_PORT', 27017),
>     'database' => env('DB_DATABASE'),
>     'username' => env('DB_USERNAME'),
>     'password' => env('DB_PASSWORD'),
>     'options'  => [
> 		'replicaSet' => 'replicaSetName'
> 	]
> ],
> ```

I'd just like to make sure that if there are any misleading documentation or code examples laying around, we can get those cleaned up and help others avoid this issue in the future.

**Comment by gmsantos** [comment]
Hi, one of the maintainers of [leroy-merlin-br/mongolid-laravel](https://github.com/leroy-merlin-br/mongolid-laravel) here.

I guess `DB_REPLICA_SET` is just a coincidence. We only set `replicaSet` on connection if [it's not null](https://github.com/leroy-merlin-br/mongolid-laravel/blob/master/src/MongolidServiceProvider.php#L117-L119) and only for [cluster configuration only](https://github.com/leroy-merlin-br/mongolid-laravel#installation).

**Comment by zahirkelloud** [comment]
I have been pulling my hair off because of a similar issue. I was sure that I have everything correct in my connection config. I spent hours trying to figure out where the issues is. It turns out that Laravel caches the configs so I had to run ```php artisan config:clear``` to clear the cached configuration. Only then I was able to make the connection to mongodb server. So for those who tried everything and didn't work, run ```php artisan config:clear``` every time you make a change in your config file.
